Updating events

clcoffey opened the pull request:
Updating Community Call images to include the date.
Updating the banner on the home page for our next Community Call

lenucksi commented on resources/learningpath/contributor/02.md:
```diff
@@ -0,0 +1,17 @@
+learning_path_article: 'contributor/02-becoming-a-contributor-article'
+---
+
+| Contributor |
```
Re the manually generated table: Do you think it would make sense to render that one based on (some more) structure in the URL mapping table?

tsadler1988 replied:
I think we might be approaching the limits of what Jekyll can do (this PR already mostly bypasses Jekyll) but I will investigate.

tsadler1988 replied:
There should be a way to create a ToC from data, but I think the section titles and subtitles (e.g. Contributor > 02 - Becoming an InnerSource Contributor) will likely need to be static. 

URLs for videos and articles can still be retrieved dynamically though.

Alternative approaches:
- Generate navigation and articles every time innersourcecommons.org is built
  - Will require quite a bit more work, especially as we'll probably want to implement a way to auto-build innersourcecommons.org on updates to the Learning Path
  - Generating at build could help solve the accessibility problem
  - More options for versioning and publishing the articles
- Add in lots of hacks around Jekyll
  - Probably a bad idea

jonico replied:
> There should be a way to create a ToC from data, but I think the section titles and subtitles (e.g. Contributor > 02 - Becoming an InnerSource Contributor) will likely need to be static.

https://github.com/marketplace/actions/toc-generator ?

rrrutledge replied:
I think that static titles and number of entries should be fine because we're not expecting that to ever change at this point.  URLs might change, so getting those in dynamically would be perfect.

For our use case we don't need any more flexibility than that.

rrrutledge replied:
Since all of the pages are numbered, you could dynamically generate "next" and "previous" links and have them at the top and bottom of each article.

tsadler1988 replied:
Investigated `toc-generator` but I beleive that generates TOCs based on headings within a markdown file, rather than a directory structure, unless I'm missing something.

https://github.com/forestryio/jekyll-menus might be a goer, however.

tsadler1988 replied:
Managed to put something together - see latest commit.

lenucksi commented:
Thanks @tsadler1988 this is great! :D 
I also like the approach of bypassing intermediate products using asciidoctor.js.

> **Issues/questions:**
> 
>  * Is raw GitHub appropriate for this? Tried using https://gitcdn.link/ but it appears to be having issues. Alternatively, could we publish from InnerSourceLearningPath to a CDN or similar ourselves?

I think GitHub should be ok with that sort of load in its current state. They do GitHub pages too which, maybe we can leverage that too. Maybe @jonico might be able to help us out with a bit of advice and  the opinion of GitHub here.

> 
>  * Navigating between articles is done via a manually generated table. Suspect there's a better way.

Maybe the structure (or an extension of it) in https://github.com/InnerSourceCommons/InnerSourceLearningPath/blob/master/config/urls.yaml could help us here?

> 
>  * Should use https://github.com/InnerSourceCommons/InnerSourceLearningPath/blob/master/config/urls.yaml to get the asciidoc URLs, rather than hardcoding (but the raw GitHub question still stands).

Yes please. @rrrutledge surely would appreciate that after having invested the effort to create this in the first place :)

jonico commented:
> Maybe @jonico might be able to help us out with a bit of advice and the opinion of GitHub here.

GitHub Pages is backed by a CDN and is GitHub’s recommend way to play out content (which could be mirrored from another source).

jonico commented:
> Navigating between articles is done via a manually generated table. Suspect there's a better way.

There are GitHub Actions to both generate a TOC as well as to publish any material on GitHub Pages. Public repos have no hard cap on Action minutes.

tsadler1988 commented on _includes/_head.html:
```diff
@@ -25,7 +25,8 @@
 	<script src="{{ '/assets/js/modernizr.min.js' | relative_url }}"></script>
-  <script src="{{ url }}/assets/js/events.js"></script>
+	<script src="{{ url }}/assets/js/events.js"></script>
+	<script src="{{ site.url }}{{ site.baseurl }}/assets/js/javascript.min.js"></script>
```
This is jQuery, and is currently being loaded in the footer under the same name.

I've changed my approach to use `fetch` rather than `$.get` and reverted this change anyway (see my latest commit) so I don't think we should consider renaming in this PR.

tsadler1988 commented on _includes/_head.html:
```diff
@@ -79,6 +80,10 @@
 	{% endif %}
 
+	{% if page.learning_path_article %}
+		<script src="{{ url }}/assets/js/asciidoctor.min.js"></script>
```
Good point @lenucksi, looks like asciidoctor.js, and asciidoctor more generally, is lacking here (see https://github.com/asciidoctor/asciidoctor/issues/242).

There are some alternative HTML5 backends for asciidoctor that might address this, but they can't be run in a browser:
- https://github.com/jirutka/asciidoctor-html5s (Ruby and Node.JS only)
- https://github.com/oreillymedia/asciidoctor-htmlbook (Ruby only)

Therefore I suggest we:
1. Don't render on the client
1. Investigate transpiling one of the alternative backends for use on browser
1. Extend asciidoctor.js ourselves to improve accessibility

If we stick with rendering on the client, I would raise this as a separate issue (with 2 or 3 as possible solutions).

rrrutledge replied:
Agree with separate issue.  This PR moves the learning path forward.  Accessibility would move it forward even further, but no need to delay what we have here.

lenucksi replied:
Entirely agree with @rrrutledge here. Let's keep this on our screen using a separate issue and move forward with this in this iteration. 
I'll see if I can get a report from a friend of mine what [JAWS](https://www.freedomscientific.com/products/software/jaws/) says about this PR's result once its merged.

tsadler1988 replied:
https://github.com/InnerSourceCommons/InnerSourceLearningPath/issues/247 raised for this.

tsadler1988 commented on _layouts/page.html:
```diff
@@ -40,6 +40,17 @@ <h1>{{ page.title }}</h1>
+			<div id="learningPathArticle"></div>
+			<script>
+				function learningPathCallback (response) {
+					$('#learningPathArticle').html(Asciidoctor().convert(response))
```
I chose master so that the website would automatically update when the articles do, without any additional publishing. But we could have a 'latest' branch on the Learning Path that we merge master into when we want to publish it, and point to that instead,

Note that if we change approach to server side rendering the asciidoc, we will need additional publishing steps anyway.

lenucksi replied:
Another option might be to go and use our .net domain for a `latest` publication at some point. Lets move forward with the `master` for this iteration! :)

rrrutledge replied:
Agree 👍

tsadler1988 commented:
Thanks @jonico. I guess that would mean creating gh-pages for https://github.com/InnerSourceCommons/InnerSourceLearningPath/, and importing fragments hosted there into here? Unless there's some other way to share gh-pages assets across projects?

That might help with some of the other issues e.g. accessibility too...

jonico commented:
> Thanks @jonico. I guess that would mean creating gh-pages for https://github.com/InnerSourceCommons/InnerSourceLearningPath/, and importing fragments hosted there into here? Unless there's some other way to share gh-pages assets across projects?
> 
> 
> 
> That might help with some of the other issues e.g. accessibility too...

You could use submodules (https://help.github.com/en/github/working-with-github-pages/using-submodules-with-github-pages), but the GitHub Actions approach is probably less complicated.

rrrutledge commented on _includes/_head.html:
```diff
@@ -79,6 +79,10 @@
 	{% endif %}
 
+	{% if page.learning_path_article %}
+		<script src="{{ url }}/assets/js/asciidoctor.min.js"></script>
```
Use a CDN for this instead of self-hosting?

https://cdnjs.com/libraries/asciidoctor.js/

tsadler1988 replied:
:+1:

rrrutledge commented on _layouts/page.html:
```diff
@@ -40,6 +40,22 @@ <h1>{{ page.title }}</h1>
 			{{ content }}
 			</div>
 
+			{% if page.learning_path_article %}
```
Would it be better practice to use [layout inheritance](https://jekyllrb.com/docs/layouts/) and have some kind of `learning-path-page` that implements this functionality?

tsadler1988 replied:
:+1:

rrrutledge commented on _layouts/page.html:
```diff
@@ -40,6 +40,22 @@ <h1>{{ page.title }}</h1>
+			<script>
+				fetch('https://raw.githubusercontent.com/InnerSourceCommons/InnerSourceLearningPath/master/{{ page.learning_path_article }}.asciidoc')
+					.then((response) => {
+						debugger
```
Can remove debugging statements before merge.

tsadler1988 replied:
Good spot!

rrrutledge replied:
Nice.

rrrutledge commented on resources/learningpath/contributor/index.md:
```diff
@@ -0,0 +1,17 @@
+
+| Contributor |
+| ------- |
+| 01 - Introduction |
```
For consistency have this file called `01.md`?

tsadler1988 replied:
I used `index.md` so we could have the path `/resources/learningpath/contributor/` but I'm not too precious about it.

Could also implement redirects so that  `/resources/learningpath/contributor/` will map to  `/resources/learningpath/contributor/01`.

tsadler1988 replied:
Had a quick attempt at redirect by implementing an `index.md`:
```
---
redirect_to:
    - 01
---
```
But that means `/resources/learningpath/contributor/01` 404s despite `01.md` existing - it is only accessible via `/resources/learningpath/contributor`. So we will have to choose one or the other.

rrrutledge commented on _layouts/page.html:
```diff
@@ -40,6 +40,22 @@ <h1>{{ page.title }}</h1>
+					})
+					.then((text) => {
+						debugger
+						document.getElementById('learningPathArticle').innerHTML = Asciidoctor().convert(text)
```
Trying to figure out how this works - how does a `#learningPathArticle` get on the pages?  I don't see it anywhere in the page MD.

Just want to understand how things work, here.

tsadler1988 replied:
See just above the `<script>` block I create an empty `<div>` with that ID.

rrrutledge replied:
I can't believe I missed it!  Thanks.

tsadler1988 commented:
Discussed in the Learning Path weekly call. We are happy to use this approach for an MVP, so I will finish what I've started including addressing comments raised (thanks everyone for the comments!).

Follow up tasks:
- Improve accessibility https://github.com/InnerSourceCommons/InnerSourceLearningPath/issues/247
- Stop using raw GitHub https://github.com/InnerSourceCommons/InnerSourceLearningPath/issues/248
- Link to self hosted rather GitHub https://github.com/InnerSourceCommons/InnerSourceLearningPath/issues/207
- Self-host non-English articles https://github.com/InnerSourceCommons/InnerSourceLearningPath/issues/249

tsadler1988 commented on _includes/_head.html:
```diff
@@ -79,6 +79,10 @@
 	{% endif %}
 
+	{% if page.learning_path_article %}
+		<script src="https://cdnjs.cloudflare.com/ajax/libs/asciidoctor.js/1.5.9/asciidoctor.min.js"></script>
```
:+1:

tsadler1988 commented:
Implemented @rrrutledge's previous/next suggestion at the bottom of the page. 

![image](https://user-images.githubusercontent.com/5602024/81242696-b5f21980-9005-11ea-85b3-87db61753ff4.png)

tsadler1988 commented:
Think this is ready to merge now, pending review of my latest commits (thanks in advance!).

clcoffey committed:
Updating events

Updating events images to include the date.
Updating the banner on the home page for our next event

claredillon reviewed (commented):
Looks good

voborgus commented:
looks amazing! thanks @clcoffey
